courses/views.py

```diff
@@ -222,9 +226,12 @@ def student_can_add_course(request, course_id, student_id):
     course = Course.objects.get(pk=course_id)
     student = Student.objects.get(pk=student_id)
     if course.student_registration_open:
-        course.students.add(student)
-        student.save()
-        messages.success(request, 'You are enrolled in %s' % (course))
+        if student not in course.students.all():
```

Comment:
Great catch on this one .. and yes .. your solution makes sense . and it should be correct

However, there is a problem in web applications when u solve it in a view function .. if you hit refresh twice quickly .. you can register twice even if you have this condition

The problem is, the delay in communication would result in the requests arriving almost at the same time .. and because both requests are done at the same time .. they might check the if statement together at the same time

Both requests will see that the student is not registered ...
and they will both register at the same time

So you might think the solution is correct ,... but the web environment creates a condition where the if statement will not work

The solution to any problem related to data integrity must be pushed to the database

So the correct solution has to happen in models.py

We add a Meta class, which is a class describing options related to the model
The meta class is placed inside Course model
and inside the meta class we set values for all the options we want
The option we want to fix is unique_together .. we tell the database that a combination of fields must be unique .. instead of a single value
this way the database checks for uniqueness across multiple values

What further complicates things is that this is a manytomany relationship .. so we must create a through table (join table) .. and we put the unique_together constraint on it.

Please read the following to know what Meta class is:
https://docs.djangoproject.com/en/1.9/topics/db/models/#meta-options

and read the following to understand the unique_together options:
https://docs.djangoproject.com/en/1.9/ref/models/options/#unique-together

I will reject this one .. but I will fix the issue and point u to view the correct solution inshallah